File: 2018/Python/day01.py
# Advent of Code 2018 - Day 1
from itertools import accumulate
from itertools import cycle


# -- Part 1
with open('data/day01.txt') as f:
    print('Part 1:', sum(int(i) for i in f))

# -- Part 2
with open('data/day01.txt') as f:
    nums = [int(i) for i in f]

seen = {0}

for i in accumulate(cycle(nums)):
    if i not in seen:
        seen.add(i)
    else:
        print('Part 2:', i)
        break

File: 2018/Haskell/src/Day01.hs
module Day01 where

import Data.Set (Set)
import qualified Data.Set as Set

parse :: String -> [Int]
parse = map read . lines . filter (/= '+')

firstRepeat :: Set Int -> [Int] -> Int
firstRepeat _ [] = undefined
firstRepeat seen (x : xs) =
  if Set.member x seen
    then x
    else firstRepeat (Set.insert x seen) xs

main :: IO ()
main = do
  input <- parse <$> readFile "../data/day01.in"
  print $ sum input
  print $ firstRepeat Set.empty . scanl (+) 0 $ cycle input

-- 533
-- 73272
